lib/spack/spack/solver/asp.py

```diff
@@ -862,6 +862,13 @@ def visit(node):
         self.control.load(os.path.join(parent_dir, "display.lp"))
         if not setup.concretize_everything:
             self.control.load(os.path.join(parent_dir, "when_possible.lp"))
+        flags = []
+        for spec in specs:
+            flags.append(spec.compiler_flags)
+
+        if self._compiler_flag_has_propagation(flags):
+            self.control.load(os.path.join(parent_dir, "propagation.lp"))
```

Comment:
There are two issues here:

1. This won't account for propagation that originates in a non-root, e.g. `spack install hdf5 ^mpich++debug`
2. This requires duplicated iteration here and in `_compiler_flag_has_propagation`, which is part of why the latter is hard to read.

What about something like:

```suggestion
        has_propagation = False
        for spec in specs:
            for dep in spec.traverse(root=True):
                has_propagation |= self._compiler_flag_has_propagation(dep.compiler_flags)

        if has_propagation:
            self.control.load(os.path.join(parent_dir, "propagation.lp"))
```

This could then work in conjunction with the simplified form of `_compiler_flag_has_propagation` that I commented on that method.